src/sql-parser/src/parser.rs

```diff
@@ -192,6 +203,193 @@ impl ParserError {
+                    Some((StatementAction::Create, AuditObjectType::ClusterReplica))
+                }
+                StatementKind::CreateConnection => {
+                    Some((StatementAction::Create, AuditObjectType::Connection))
+                }
+                StatementKind::CreateDatabase => {
+                    Some((StatementAction::Create, AuditObjectType::Database))
+                }
+                StatementKind::CreateIndex => {
+                    Some((StatementAction::Create, AuditObjectType::Index))
+                }
+                StatementKind::CreateMaterializedView => {
+                    Some((StatementAction::Create, AuditObjectType::MaterializedView))
+                }
+                StatementKind::CreateRole => Some((StatementAction::Create, AuditObjectType::Role)),
+                StatementKind::CreateSchema => {
+                    Some((StatementAction::Create, AuditObjectType::Schema))
+                }
+                StatementKind::CreateSecret => {
+                    Some((StatementAction::Create, AuditObjectType::Secret))
+                }
+                StatementKind::CreateSink => Some((StatementAction::Create, AuditObjectType::Sink)),
+                StatementKind::CreateSource => {
+                    Some((StatementAction::Create, AuditObjectType::Source))
+                }
+                StatementKind::CreateSubsource => {
+                    Some((StatementAction::Create, AuditObjectType::Subsource))
+                }
+                StatementKind::CreateTable => {
+                    Some((StatementAction::Create, AuditObjectType::Table))
+                }
+                StatementKind::CreateView => Some((StatementAction::Create, AuditObjectType::View)),
+                _ => None,
+            })
+            .flatten()
+    }
+}
+
+impl fmt::Display for ParserStatementError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str(&format!("{}: {}", &self.statement_type(), &self.error))
+    }
+}
+
+pub trait ParserStatementErrorMapper<T> {
+    fn map_parser_err(self, statement_kind: StatementKind) -> Result<T, ParserStatementError>;
+
+    fn map_no_statement_parser_err(self) -> Result<T, ParserStatementError>;
+}
+
+impl<T> ParserStatementErrorMapper<T> for Result<T, ParserError> {
+    fn map_parser_err(self, statement: StatementKind) -> Result<T, ParserStatementError> {
+        self.map_err(|error| ParserStatementError {
+            error,
+            statement: Some(statement),
+        })
+    }
+
+    fn map_no_statement_parser_err(self) -> Result<T, ParserStatementError> {
+        self.map_err(|error| ParserStatementError {
+            error,
+            statement: None,
+        })
+    }
+}
+
+impl From<ParserError> for ParserStatementError {
+    fn from(error: ParserError) -> Self {
+        Self {
+            error,
+            statement: None,
+        }
+    }
+}
```

Comment:
This `From` implementation looks like it hides a lot of cases where we _should_ be adding a `map_parser_err()` call because the `StatementKind` is known.